portable/GCC/ATmegaxxxx/portmacro.h

```diff
@@ -0,0 +1,149 @@
+#endif
+/*-----------------------------------------------------------*/
+
+/* Critical section management. */
+
+#define portENTER_CRITICAL()        __asm__ __volatile__ (                          \
+                                        "in __tmp_reg__, __SREG__"        "\n\t"    \
+                                        "cli"                             "\n\t"    \
+                                        "push __tmp_reg__"                "\n\t"    \
+                                        ::: "memory"                                \
+                                        )
+
+
+#define portEXIT_CRITICAL()         __asm__ __volatile__ (                          \
+                                        "pop __tmp_reg__"                 "\n\t"    \
+                                        "out __SREG__, __tmp_reg__"       "\n\t"    \
+                                        ::: "memory"                                \
+                                        )
+
+
+#define portDISABLE_INTERRUPTS()    __asm__ __volatile__ ( "cli" ::: "memory")
+#define portENABLE_INTERRUPTS()     __asm__ __volatile__ ( "sei" ::: "memory")
+
+/*-----------------------------------------------------------*/
+
+/* Architecture specifics. */
+
+// System Tick  - Scheduler timer
+// Prefer to use the enhanced Watchdog Timer, but also Timer0 is ok.
+
+#if defined(WDIE) && defined(WDIF)              // Enhanced WDT with interrupt capability
+
+#define portUSE_WDTO                WDTO_15MS   // portUSE_WDTO to use the Watchdog Timer for xTaskIncrementTick
+
+/* Watchdog period options:         WDTO_15MS
+                                    WDTO_30MS
+                                    WDTO_60MS
+                                    WDTO_120MS
+                                    WDTO_250MS
+                                    WDTO_500MS
+                                    WDTO_1S
+                                    WDTO_2S
+*/
+
+#else
+
+#define portUSE_TIMER0                          // portUSE_TIMER0 to use 8 bit Timer0 for xTaskIncrementTick
+
+#endif
+
+#define portSTACK_GROWTH            ( -1 )
+
+/* Timing for the scheduler.
+ * Watchdog Timer is 128kHz nominal,
+ * but 120 kHz at 5V DC and 25 degrees is actually more accurate,
+ * from data sheet.
+ */
+#if defined( portUSE_WDTO )
+#define portTICK_PERIOD_MS          ( (TickType_t) _BV( portUSE_WDTO + 4 ) )    // Inaccurately assuming 128 kHz Watchdog Timer.
+#else
+#define portTICK_PERIOD_MS          ( ( TickType_t ) 1000 / configTICK_RATE_HZ )
+#endif
+
+#define portBYTE_ALIGNMENT          1
+#define portNOP()                   __asm__ __volatile__ ( "nop" );
+/*-----------------------------------------------------------*/
+
+/* Kernel utilities. */
+extern void vPortYield( void )      __attribute__ ( ( naked ) );
+#define portYIELD()                 vPortYield()
+/*-----------------------------------------------------------*/
+
+#if defined(__AVR_ATmega2560__) || defined(__AVR_ATmega2561__)
```

Comment:
I presume these parts are singled out (doubled out?) because they have larger memory space.  In general preference is not to have chip (as opposed to architecture) specifics in a port layer, especially if it puts a burden on the user to get the compile time options right, or the target chip family is likely to change in the future.  So a couple of questions - is there a catch-all definition that can be use for all devices for which this special use case applies, or alternatively, can we just have this special use case apply to all chips no matter if they need it or not?    [either way, the code as you have it now is better than the legacy code, which was very chip family specific.]

Is .lowtext something that always exists, or something the user will potentially trip up on if they don't have the correct linker script.  If the latter is true then please update the comment to tell users how to fix any potential linker errors.  (can just be a link to a forum post somewhere that tells people how to fix).

Reply:
The two parts __ATmega2560__ and __ATmega2561__ are the only two parts in common usage (and for the past 12 years at least) that need the 3 Byte PC solution.

Potentially they could also be differentiated by architecture by using the [these options](https://gcc.gnu.org/onlinedocs/gcc/AVR-Options.html) and providing the below.

```
#if defined(__AVR_3_BYTE_PC__)
```
This is more generic, so I'll fix it.

The `.lowtext` section is provided by the linker script in `avr6.x`. If they have correctly nominated their device type in the compile line `-mmcu=atmegaxxxx`, then the correct linker script will be selected. It is automatic.

Reply:
There is an additional twist to this story in that one device has both 256kB of Flash and needs `__AVR_3_BYTE_PC__` and can address XRAM beyond 64kB and so has the `__AVR_HAVE_RAMPZ__` register. This is the ATmega2560.

To cater for the ATmega2561, which has a 3 Byte PC but doesn't have XRAM capability, another option for `portSAVE_CONTEXT()` needs to be provided.

__EDIT__ done.